napari/layers/labels/labels.py

```diff
@@ -562,15 +561,34 @@ def _set_colormap(self, colormap):
     @property
     def num_colors(self):
         """int: Number of unique colors to use in colormap."""
-        return self._num_colors
+        warnings.warn(
+            trans._(
+                'Labels.num_colors is deprecated since 0.4.19 and will be '
+                'removed in 0.5.0, please use len(Labels.colormap) - 1 '
+                'instead.',
+                deferred=True,
+            ),
+            FutureWarning,
+            stacklevel=2,
+        )
```

Comment:
```suggestion
        warnings.warn(
            trans._(
                'Labels.num_colors is deprecated since 0.4.19 and will be '
                'removed in 0.5.0, please use Labels.colormap.num_colors '
                'instead.',
                deferred=True,
            ),
            FutureWarning,
            stacklevel=2,
        )
```

Reply:
I do not like this -1. It is technical detail. We should define property (make `_num_unique_colors` property public.

Reply:
Actually, the -1 is outdated — if you look below, there's no -1. I was just confused because we use different conventions in different places. I think in the future we should harmonize on one definition or the other (I'd say probably including all colors, since then it matches the length of the colorarray) and keep len, which is intuitive. For now I will just update the message to just `len(Label.colormap)` (no -1).

Reply:
```python
len(viewer.layers[0].colormap)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 len(viewer.layers[0].colormap)

TypeError: object of type 'LabelColormap' has no len()
```

Reply:
Yes you need this PR for that 😉

Reply:
It will not work with direct colormap.

Reply:
Interesting point! I'll think about ways around that... It would be easy to add a len to the direct colormap but then it is a behaviour change. But it's also a bit weird to ask for num_colors while in direct mode. 🤔

Reply:
And all this problems comes from joining both colormap under one name. 

Maybe t is better to deprecate `colormap` property in Labels layer, and introduce `cyclic_colormap` and `direct_colormap` ones with keeping `color_mode` property?

Reply:
Question: would you expect len(a_direct_colormap) to give you the number of colors or the number of keys? Colors is of course much easier to implement and I'm inclined to just do that. But I'm willing to hear counter-arguments. 😊

Reply:
> Keys is much easier to implement (`len(dict)`) than colors (`len(set(tuple(x) for x in dict.values()))`)

@Czaki no, the point is that actually, it's already implemented for the values, because I added `len` to `Colormap`, not even `LabelsColormapBase`, as `len(self.colors)`. Since you construct `DirectLabelColormap` by making the above set, it already returns the number of colors. Since the answer is accurate (this is in fact the number of colors), I'm inclined to just leave it as-is... For example, if one does a mapping of 1M labels to Viridis discretised to uint8, I think 256 is a reasonable answer.

I guess the question is why people would be reading `layer.num_colors` if they are using direct mode... My guess is that this doesn't happen. (And if it does, we will hear about it as a bug report during the deprecation period (0.4.19 -> 0.5.0).)

Reply:
(Hmmm, so do you think I should return `len(self._random_colormap)` in the deprecated implementation? Then people can notice that the results are different when they implement the suggestion.)

Reply:
> I guess the question is why people would be reading `layer.num_colors` if they are using direct mode... My guess is that this doesn't happen. (And if it does, we will hear about it as a bug report during the deprecation period (0.4.19 -> 0.5.0).)

Because some plugin creators will not think earlier and forget that under colormap it could also be direct one. 

> Since you construct `DirectLabelColormap` by making the above set, it already returns the number of colors

`colors` in `DirecLabelColorap` will have length 1.

https://github.com/napari/napari/blob/d507dd64ed6d96e984fa7fa57f2f232be388b5f7/napari/utils/colormaps/colormap.py#L353-L356

In a direct colormap the colors are stored in `color_dict` attribute. 

It i not my design. I just modified the existing one. 

> do you think I should return `len(self._random_colormap)` in the deprecated implementation

For me, it will be a number of keys. But without strong opinion.

Reply:
@Czaki I went with passing through `_num_unique_colors` because it allows unifying a few code paths between direct and auto color modes.